geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/KryoLazyDensityFilter.java

```diff
@@ -0,0 +1,379 @@
+        return (gd != null && gd.getType().getBinding() == Point.class);
+    }
+
+    Map.Entry<Tuple2<Integer, Integer>, Double> writePoint(SimpleFeature sf, Double weight) {
+        return writePointToResult((Point) sf.getAttribute(geomIndex), weight);
+    }
+
+    private Map.Entry<Tuple2<Integer, Integer>, Double> writeNonPoint(Geometry geom, Double weight) {
+        return writePointToResult(safeCentroid(geom), weight);
+    }
+
+    private Point safeCentroid(Geometry geom) {
+        Polygon p = geom.noPolygon();
+        Point centroid = p.getCentroid();
+        if (Double.isNaN(centroid.getCoordinate().x) || Double.isNaN(centroid.getCoordinate().y)) {
+            return p.getEnvelope().getCentroid();
+        } else {
+            return centroid;
+        }
+    }
+
+    private Map.Entry<Tuple2<Integer, Integer>, Double> writePointToResult(Point pt, Double weight) {
+        return writeSnappedPoint(gridSnap.i(pt.getX()), gridSnap.j(pt.getY()), weight);
+    }
+
+    private Map.Entry<Tuple2<Integer, Integer>, Double> writeSnappedPoint(int x, int y, Double weight) {
+        Tuple2 tuple = new Tuple2(x, y);
+        return new AbstractMap.SimpleEntry(tuple, weight);
+
+    }
+
+    private void writePointToResult(Coordinate pt, Double weight) {
+        writeSnappedPoint(gridSnap.i(pt.x), gridSnap.j(pt.y), weight);
+    }
+
+    private void writePointToResult(double x, double y, double weight) {
+        writeSnappedPoint(gridSnap.i(x), gridSnap.j(y), weight);
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell cell) throws IOException {
+        return ReturnCode.INCLUDE;
+    }
+
+    @Override
+    public Cell transformCell(Cell cell) throws IOException {
+        byte[] row = CellUtil.cloneRow(cell);
+        SimpleFeature sf_old = serializer.deserialize(cell.getValueArray(), cell.getValueOffset(), cell.getValueLength());
+        Map.Entry<Tuple2<Integer, Integer>, Double> out;
+        if (isPoints()) {
+            out = writePoint(sf_old, weightFn(sf_old));
+        } else {
+            out = writeNonPoint((Geometry) sf_old.getDefaultGeometry(), weightFn(sf_old));
+        }
+        SimpleFeature sf = new ScalaSimpleFeature(sf_old.getID(), outputsft, null, null);
+        sf.setAttribute(0, serializeParameters(out.getKey()));
+        sf.setAttribute(1, out.getValue());
+        byte[] arr = output_serializer.serialize(sf);
+        Cell newCell = CellUtil.createCell(row, arr);
+        return newCell;
+    }
+
+    // TODO: Add static method to compute byte array from SFT and Filter.
+    @Override
+    public byte[] toByteArray() throws IOException {
+        return Bytes.add(getLengthArray(sftString), serializeHashMap(options));
+    }
+
+    private byte[] serializeHashMap(Map<String, Object> map) throws IOException{
+        ByteArrayOutputStream fis = new ByteArrayOutputStream();
+        ObjectOutputStream ois = new ObjectOutputStream(fis);
+        ois.writeObject(map);
```

Comment:
we really want to avoid java serialization if possible... the density options are encodable as just key/value strings. adding a generalized `Map[String,String]` serialization/deserialization would allow us to use the same configs in accumulo and hbase.